### PR TITLE
[FLINK-1219] Add support for Apache Tez as execution backend

### PR DESCRIPTION
**Pull request by ktzoumas**
This allows Flink programs to be executed as Tez DAGs.

Up for comments, not good to merge yet.

The commit to make hadoop2 the default profile will eventually be a separate commit and issue,

### CODE REVIEW

**Comment by ktzoumas** [comment]
I tested this on a 25-node cluster with WordCount and the modified TPC-H query on generated TPC-H data of scale 1000. I also added docs.

File: flink-staging/flink-tez/pom.xml [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+            <artifactId>commons-codec</artifactId>
+            <version>1.4</version>
+        </dependency>
+
+        <!--  guava needs to be in "provided" scope, to make sure it is not included into the jars by the shading -->
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>11.0.2</version>
+            <scope>compile</scope>
```
Comment (rmetzger) [inline flink-staging/flink-tez/pom.xml]: we need to have "provided" scope here until my shading PR is ready.

File: flink-staging/flink-tez/pom.xml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                        <phase>package</phase>
+                        <goals>
+                            <goal>single</goal>
+                        </goals>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.rat</groupId>
+                <artifactId>apache-rat-plugin</artifactId>
```
Comment (rmetzger) [inline flink-staging/flink-tez/pom.xml]: is this disabling rat?

File: flink-staging/flink-tez/src/main/java/org/apache/flink/tez/client/RemoteTezEnvironment.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	private static final Log LOG = LogFactory.getLog(TezExecutor.class);
+	
+	private PactCompiler compiler;
+	private TezExecutor executor;
+	private Path jarPath = null;
+	
+
+	public void registerMainClass (Class mainClass) {
+		jarPath = new Path(ClassUtil.findContainingJar(mainClass));
+		//this.jarURL = mainClass.getProtectionDomain().getCodeSource().getLocation();
```
Comment (rmetzger) [inline flink-staging/flink-tez/src/main/java/org/apache/flink/tez/client/RemoteTezEnvironment.java]: this looks like debugging code

File: flink-staging/flink-tez/src/main/java/org/apache/flink/tez/client/RemoteTezEnvironment.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.flink.compiler.costs.DefaultCostEstimator;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.util.ClassUtil;
+import org.apache.hadoop.util.ToolRunner;
+
+
+public class RemoteTezEnvironment extends ExecutionEnvironment {
+
+	private static final Log LOG = LogFactory.getLog(TezExecutor.class);
```
Comment (rmetzger) [inline flink-staging/flink-tez/src/main/java/org/apache/flink/tez/client/RemoteTezEnvironment.java]: The log messages will appear from the wrong class.

File: docs/flink_on_tez_guide.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+  add the following entry to tez-site.xml
+    ~~~<property>
+      <name>tez.aux.uris</name>
+      <value>${fs.default.name}/apps/flink-tez-x.y.z-flink-fat-jar.jar</value>
+    </property>
+    ~~~
+    
+- At this point, you should be able to run the pre-packaged examples, e.g., run WordCount as:
+  {% highlight bash %}
+  $ hadoop jar /path/to/flink-tez-x.y.z-flink-fat-jar.jar org.apache.flink.tez.ExampleDriver \
```
Comment (rmetzger) [inline docs/flink_on_tez_guide.md]: The classname `org.apache.flink.tez.ExampleDriver` is wrong.

**Comment by rmetzger** [comment]
The documentation is very well written.

I've tried the pull request on a Hortonworks Sandbox with HDP 2.2.

They have Tez 0.5.2.2.2.0.0 installed there. The mvn command to build Tez for this environment is:
```
mvn install -DskipTests -Pinclude-tez -Dhadoop.version=2.6.0 -Dtez.version=0.5.2.2.2.0.0-2041 -Pvendor-repos
```

To change the tez version, I made the following changes to the pom:
```diff
diff --git a/flink-staging/flink-tez/pom.xml b/flink-staging/flink-tez/pom.xml
index e84f731..147dd32 100644
--- a/flink-staging/flink-tez/pom.xml
+++ b/flink-staging/flink-tez/pom.xml
@@ -34,6 +34,9 @@ under the License.
     <name>flink-tez</name>
 
     <packaging>jar</packaging>
+    <properties>
+       <tez.version>0.6.0</tez.version>
+    </properties>
 
     <dependencies>
         <dependency>
@@ -81,25 +84,25 @@ under the License.
         <dependency>
             <groupId>org.apache.tez</groupId>
             <artifactId>tez-api</artifactId>
-            <version>0.6.0</version>
+            <version>${tez.version}</version>
         </dependency>
 
         <dependency>
             <groupId>org.apache.tez</groupId>
             <artifactId>tez-common</artifactId>
-            <version>0.6.0</version>
+            <version>${tez.version}</version>
         </dependency>
 
         <dependency>
             <groupId>org.apache.tez</groupId>
             <artifactId>tez-dag</artifactId>
-            <version>0.6.0</version>
+            <version>${tez.version}</version>
         </dependency>
 
         <dependency>
             <groupId>org.apache.tez</groupId>
             <artifactId>tez-runtime-library</artifactId>
-            <version>0.6.0</version>
+            <version>${tez.version}</version>
         </dependency>
 
         <dependency>
```

Once these issues have been resolved, I'd say the change is +1 to merge.

**Comment by rmetzger** [comment]
Fix for the build failure is being tested: https://travis-ci.org/rmetzger/flink/builds/56885481

**Comment by rmetzger** [comment]
I'll merge this as soon as my final travis build goes through and Apache's git repo is working again.

**Pull request closed without merging**